x/tally/keeper/gas_calculation.go

```diff
@@ -0,0 +1,222 @@
+package keeper
+
+import (
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"strconv"
+
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/sedaprotocol/seda-chain/x/tally/types"
+)
+
+// GasCalculation stores the results of the canonical gas consumption calculations.
+type GasCalculation struct {
+	Burn          uint64 // filter and tally gas to be burned
+	Proxies       []ProxyGasUsed
+	Executors     []ExecutorGasUsed
+	ReducedPayout bool
+}
+
+type ProxyGasUsed struct {
+	PublicKey []byte
+	Amount    math.Int
+}
+
+type ExecutorGasUsed struct {
+	PublicKey string
+	Amount    math.Int
+}
+
+// DistributionsFromGasCalculation constructs a list of distribution messages
+// to be sent to the core contract based on the given gas calculation.
+func (k Keeper) DistributionsFromGasCalculation(ctx sdk.Context, reqID string, gasCalc GasCalculation, gasPrice math.Int, burnRatio math.LegacyDec) []types.Distribution {
+	dists := []types.Distribution{}
+	attrs := []sdk.Attribute{
+		sdk.NewAttribute(types.AttributeDataRequestID, reqID),
+		sdk.NewAttribute(types.AttributeReducedPayout, strconv.FormatBool(gasCalc.ReducedPayout)),
+	}
+
+	// First distribution message is the combined burn.
+	burn := types.NewBurn(math.NewIntFromUint64(gasCalc.Burn), gasPrice)
+	dists = append(dists, burn)
+	attrs = append(attrs, sdk.NewAttribute(types.AttributeTallyGas, strconv.FormatUint(gasCalc.Burn, 10)))
+
+	// Append distribution messages for data proxies.
+	for _, proxy := range gasCalc.Proxies {
+		proxyDist := types.NewDataProxyReward(proxy.PublicKey, proxy.Amount, gasPrice)
+		dists = append(dists, proxyDist)
+		attrs = append(attrs, sdk.NewAttribute(types.AttributeDataProxyGas,
+			fmt.Sprintf("%s,%s", hex.EncodeToString(proxy.PublicKey), proxy.Amount.String())))
+	}
+
+	// Append distribution messages for executors, burning a portion of their
+	// payouts in case of a reduced payout scenario.
+	addBurn := math.ZeroInt()
+	for _, executor := range gasCalc.Executors {
+		payoutAmt := executor.Amount
+		if gasCalc.ReducedPayout {
+			burnAmt := burnRatio.MulInt(executor.Amount).TruncateInt()
+			payoutAmt = executor.Amount.Sub(burnAmt)
+			addBurn = addBurn.Add(burnAmt.Mul(gasPrice))
+		}
+
+		executorDist := types.NewExecutorReward(executor.PublicKey, payoutAmt, gasPrice)
+		dists = append(dists, executorDist)
+		attrs = append(attrs, sdk.NewAttribute(types.AttributeExecutorGas,
+			fmt.Sprintf("%s,%s", executor.PublicKey, payoutAmt.String())))
+	}
+
+	// Add additional burn.
+	dists[0].Burn.Amount = dists[0].Burn.Amount.Add(addBurn.Mul(gasPrice))
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeGasCalculation, attrs...))
```

Comment:
If possible I'd like to have the additional burn amount as a separate attribute in the events.

Reply:
added a new event attribute "reduced_payout_burn"